pallets/stable-staking/src/lib.rs

```diff
@@ -0,0 +1,888 @@
+
+#[frame_support::pallet]
+pub mod pallet {
+	use frame_support::transactional;
+
+	use super::*;
+
+	type BalanceOf<T> =
+		<<T as Config>::Fungibles as FsInspect<<T as frame_system::Config>::AccountId>>::Balance;
+	type NativeBalanceOf<T> =
+		<<T as Config>::Fungible as FInspect<<T as frame_system::Config>::AccountId>>::Balance;
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_assets::Config {
+		/// Overarching event type
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Origin used to administer the pallet
+		type StakingPoolCommitteeOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// Identifier for the class of pool.
+		type PoolId: Member
+			+ Parameter
+			+ Clone
+			+ MaybeSerializeDeserialize
+			+ MaxEncodedLen
+			+ Default;
+
+		type Fungibles: FsMutate<Self::AccountId>;
+
+		type Fungible: FMutate<Self::AccountId>;
+
+		/// staking account to receive assets of notion
+		#[pallet::constant]
+		type StableTokenBeneficiaryId: Get<PalletId>;
+
+		/// The beneficiary PalletId, used for deriving its sovereign AccountId for providing native
+		/// token reward
+		#[pallet::constant]
+		type NativeTokenBeneficiaryId: Get<PalletId>;
+
+		/// The maximum length of a pool name or short description stored on-chain.
+		#[pallet::constant]
+		type PoolStringLimit: Get<u32>;
+	}
+
+	// Metadata of staking pools
+	#[pallet::storage]
+	#[pallet::getter(fn staking_pool_metadata)]
+	pub type StakingPoolMetadata<T: Config> = StorageMap<
+		_,
+		Twox64Concat,
+		T::PoolId,
+		PoolMetadata<BoundedVec<u8, T::PoolStringLimit>>,
+		OptionQuery,
+	>;
+
+	// Setting of staking pools
+	#[pallet::storage]
+	#[pallet::getter(fn staking_pool_setting)]
+	pub type StakingPoolSetting<T: Config> = StorageMap<
+		_,
+		Twox64Concat,
+		T::PoolId,
+		PoolSetting<BlockNumberFor<T>, BalanceOf<T>>,
+		OptionQuery,
+	>;
+
+	// staking pools' stable token reward waiting claiming
+	// Pool id => unclaimed reward part
+	#[pallet::storage]
+	#[pallet::getter(fn stable_staking_pool_reward)]
+	pub type StableStakingPoolReward<T: Config> =
+		StorageMap<_, Twox64Concat, T::PoolId, BalanceOf<T>, ValueQuery>;
+
+	// staking pools' stable token reward waiting claiming
+	// Pool id, epcoh index => unclaimed reward part
+	// TODO: This part is not used and only for recording purpose
+	// Currently all reward if user did not claimed on time will be mixed equally based on staking
+	// weight That means if APY is low may leads to user not claiming their reward on purpose
+	#[pallet::storage]
+	#[pallet::getter(fn stable_staking_pool_epoch_reward)]
+	pub type StableStakingPoolEpochReward<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		T::PoolId,
+		Twox64Concat,
+		u128,
```

Comment:
Let me think how to improve in the next version perhaps.
The reason I do not want to use dynamic type is because currently, there are too many type conversions when calculating rewards/weight, all different types (block number, native token balance, pallet_asset balance, epoch..)

And need time to clean up the best solution of design.
For convenience, I force some of rather fixed type into u128 so no extra trait require specified.

Reply:
I see thanks